File: PetFinder_iOS/PetFinderApp/Entities/Organization.swift
//
//  Organization.swift
//  PetFinderApp
//
//  Created by user240565 on 20/11/23.
//

import Foundation


// MARK: - OrganizationsResponse
struct OrganizationsResponse: Codable {
    let organizations: [Organization]
}

// MARK: - OrganizationResponse
struct OrganizationResponse: Codable {
    let organization: Organization
}


// MARK: - Organization
struct Organization: Codable {
    let id: String
    let name: String
    let email, phone: String?
    let address: Address?
    let hours: Hours
    let url: String
    let missionStatement: String?
    let photos: [Photo]?

    enum CodingKeys: String, CodingKey {
        case id, name, email, phone, address, hours, url
        case missionStatement = "mission_statement"
        case photos
    }
    
    static var example: Organization {
        return Organization(id: "NJ333", name: "NJ333", email: "email@email.com", phone: "555-555-555", address: Address(address1: "Calle 13", address2: "Calle 12 ", city: "Almería", state: "", postcode: "", country: ""), hours: Hours(monday: "12:00", tuesday: "11:00 - 12:00", wednesday: "11:00 - 12:00", thursday: "11:00 - 12:00", friday: "", saturday: "", sunday: "11:00 - 12:00"), url: "", missionStatement: "Lorem Ipsum es simplemente el texto de relleno de las imprentas y archivos de texto. Lorem Ipsum ha sido el texto de relleno estándar de las industrias desde el año 1500, cuando un impresor (N. del T. persona que se dedica a la imprenta) desconocido usó una galería de textos y los mezcló de tal manera que logró hacer un libro de textos especimen Lorem Ipsum es simplemente el texto de relleno de las imprentas y archivos de texto. Lorem Ipsum ha sido el texto de relleno estándar de las industrias desde el año 1500, cuando un impresor (N. del T. persona que se dedica a la imprenta) desconocido usó una galería de textos y los mezcló de tal manera que logró hacer un libro de textos especimen v Lorem Ipsum es simplemente el texto de relleno de las imprentas y archivos de texto. Lorem Ipsum ha sido el texto de relleno estándar de las industrias desde el año 1500, cuando un impresor (N. del T. persona que se dedica a la imprenta) desconocido usó una galería de textos y los mezcló de tal manera que logró hacer un libro de textos especimen Lorem Ipsum es simplemente el texto de relleno de las imprentas y archivos de texto. Lorem Ipsum ha sido el texto de relleno estándar de las industrias desde el año 1500, cuando un impresor (N. del T. persona que se dedica a la imprenta) desconocido usó una galería de textos y los mezcló de tal manera que logró hacer un libro de textos especimen Lorem Ipsum es simplemente el texto de relleno de las imprentas y archivos de texto. Lorem Ipsum ha sido el texto de relleno estándar de las industrias desde el año 1500, cuando un impresor (N. del T. persona que se dedica a la imprenta) desconocido usó una galería de textos y los mezcló de tal manera que logró hacer un libro de textos especimen", photos: [Photo(small: "", medium: "", large: "", full: "")])
    }
    
    var hoursList: [Hour] {
        [
            Hour(id: 1, title: hours.monday ?? "--", label: "monday"),
            Hour(id: 2, title: hours.tuesday ?? "--", label: "tuesday"),
            Hour(id: 3, title: hours.wednesday ?? "--", label: "wednesday"),
            Hour(id: 4, title: hours.thursday ?? "--", label: "thursday"),
            Hour(id: 5, title: hours.friday ?? "--", label: "friday"),
            Hour(id: 6, title: hours.saturday ?? "--", label: "saturday"),
            Hour(id: 7, title: hours.sunday ?? "--", label: "sunday")
        ]
      }
}

// MARK: - Hours
struct Hours: Codable {
    let monday, tuesday, wednesday, thursday: String?
    let friday, saturday, sunday: String?
}

struct Hour: Identifiable {
    var id: Int
    var title: String
    var label: String
}

File: PetFinder_Flutter/pet_adopt_app/pet_adopt_app/lib/model/organization/organization.dart
import 'package:pet_adopt_app/model/common_class.dart';

class OrganizationsNetworkResponse {
  List<Organization> organizations;
  Pagination pagination;

  OrganizationsNetworkResponse({
    required this.organizations,
    required this.pagination,
  });
}

class OrganizationNetworkResponse {
  Organization organization;

  OrganizationNetworkResponse({required this.organization});
}

class Organization {
  String id;
  String name;
  String? email;
  String? phone;
  Address? address;
  Hours? hours;
  String url;
  String? missionStatement;
  List<Photo>? photos;

  Organization({
    required this.id,
    required this.name,
    required this.email,
    required this.phone,
    required this.address,
    required this.hours,
    required this.url,
    required this.missionStatement,
    required this.photos,
  });

  List<Hour> get hoursList {
    return [
      Hour(id: 1, title: "monday", label: hours?.monday ?? "--"),
      Hour(id: 2, title: "tuesday", label: hours?.tuesday ?? "--"),
      Hour(id: 3, title: "wednesday", label: hours?.wednesday ?? "--"),
      Hour(id: 4, title: "thursday", label: hours?.thursday ?? "--"),
      Hour(id: 5, title: "friday", label: hours?.friday ?? "--"),
      Hour(id: 6, title: "saturday", label: hours?.saturday ?? "--"),
      Hour(id: 7, title: "sunday", label: hours?.sunday ?? "--"),
    ];
  }

  factory Organization.fromMap(Map<String, dynamic> json) => Organization(
        id: json["id"],
        name: json["name"],
        email: json["email"],
        phone: json["phone"],
        address:
            json["address"] != null ? Address.fromMap(json["address"]) : null,
        hours: json["hours"] != null ? Hours.fromMap(json["hours"]) : null,
        url: json["url"],
        missionStatement: json["mission_statement"],
        photos: (json["photos"] as List<dynamic>?)
            ?.map((x) => Photo.fromMap(x))
            .toList(),
      );

  Map<String, dynamic> toMap() => {
        "id": id,
        "name": name,
        "email": email,
        "phone": phone,
        "address": address?.toMap(),
        "hours": hours?.toMap(),
        "url": url,
        "mission_statement": missionStatement,
        "photos": photos?.map((x) => x.toMap()),
      };
}

class Hours {
  String? monday;
  String? tuesday;
  String? wednesday;
  String? thursday;
  String? friday;
  String? saturday;
  String? sunday;

  Hours({
    required this.monday,
    required this.tuesday,
    required this.wednesday,
    required this.thursday,
    required this.friday,
    required this.saturday,
    required this.sunday,
  });

  factory Hours.fromMap(Map<String, dynamic> json) => Hours(
        monday: json["monday"],
        tuesday: json["tuesday"],
        wednesday: json["wednesday"],
        thursday: json["thursday"],
        friday: json["friday"],
        saturday: json["saturday"],
        sunday: json["sunday"],
      );

  Map<String, dynamic> toMap() => {
        "monday": monday,
        "tuesday": tuesday,
        "wednesday": wednesday,
        "thursday": thursday,
        "friday": friday,
        "saturday": saturday,
        "sunday": sunday,
      };
}

class Hour {
  final int id;
  final String title;
  final String label;

  Hour({required this.id, required this.title, required this.label});
}
